### PR TITLE
Added representation for multi-segment buffers

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
ReadOnlyBytes can represent bytes in arbitrarily segmented buffers. It's similar to ReadableBuffer, but it is 30% smaller at the cost of some operations being more expensive. 
 
@davidfowl, @pakrym 

CC: @joshfree, @shiftylogic, @ahsonkhan,

### CODE REVIEW

File: src/System.Buffers.Experimental/System/Buffers/ArraySegmentExtensions.cs [hunk 1]
```diff
@@ -2,26 +2,124 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Collections.Sequences;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
 
 namespace System.Buffers
 {
     public static class VariousExtensions
```
Comment (davidfowl) [inline src/System.Buffers.Experimental/System/Buffers/ArraySegmentExtensions.cs]: This name is pretty funny. Make it more specific or split up the extension classes

File: src/System.Buffers.Experimental/System/Buffers/ArraySegmentExtensions.cs [hunk 2]
```diff
@@ -2,26 +2,124 @@
...
+
+            // we only need to search the end of the first buffer. More precisely, only up to value.Length - 1 bytes in the first buffer
+            // The other bytes in first, were already search and presumably did not match
+            int bytesToSkipFromFirst = 0; 
+            if (first.Length > value.Length - 1)
+            {
+                bytesToSkipFromFirst = first.Length - value.Length - 1;
+            }
+
+            // not that we know how many bytes we need to skip, create slice of first buffer with bytes that need to be searched.
```
Comment (davidfowl) [inline src/System.Buffers.Experimental/System/Buffers/ArraySegmentExtensions.cs]: now*

File: src/System.Buffers.Experimental/System/Buffers/ArraySegmentExtensions.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/System/Buffers/ArraySegmentExtensions.cs]: is funny bad? :-)

Renamed to BufferExtensions

File: src/System.Buffers.Experimental/System/Buffers/ArraySegmentExtensions.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/System/Buffers/ArraySegmentExtensions.cs]: fixed

**Pull request merged**